runtime/src/accounts_db.rs

```diff
@@ -931,6 +960,100 @@ impl AccountsDB {
+    }
+
+    fn collect_uncleaned_pubkeys_to_slot(&self, max_slot: Slot) -> (Vec<Vec<Pubkey>>, Slot) {
+        let mut max_slot_in_uncleaned_pubkeys = 0;
+        let slots: Vec<Slot> = self
+            .uncleaned_pubkeys
+            .iter()
+            .filter_map(|entry| {
+                let slot = entry.key();
+                max_slot_in_uncleaned_pubkeys = max_slot_in_uncleaned_pubkeys.max(*slot);
+                if *slot <= max_slot {
+                    Some(*slot)
+                } else {
+                    None
+                }
+            })
+            .collect();
+        (
+            slots
+                .into_iter()
+                .filter_map(|slot| {
+                    let maybe_slot_keys = self.uncleaned_pubkeys.remove(&slot);
+                    if self.accounts_index.is_root(slot) {
+                        // Safe to unwrap on rooted slots since this is called from clean_accounts
+                        // and only clean_accounts operates on rooted slots. purge_slots only
+                        // operates on uncleaned_pubkeys
+                        let (_slot, keys) = maybe_slot_keys.expect("Root slot should exist");
+                        Some(keys)
+                    } else {
+                        None
+                    }
+                })
+                .collect(),
+            max_slot_in_uncleaned_pubkeys,
+        )
+    }
+
+    // Construct a vec of pubkeys for cleaning from:
+    //   uncleaned_pubkeys - the delta set of updated pubkeys in rooted slots from the last clean
+    //   zero_lamport_pubkeys - set of all alive pubkeys containing 0-lamport updates
+    fn construct_candidate_clean_keys(
+        &self,
+        max_clean_root: Option<Slot>,
+        timings: &mut CleanKeyTimings,
+    ) -> Vec<Pubkey> {
+        let mut zero_lamport_key_clone = Measure::start("zero_lamport_key");
+        let mut pubkeys = self.accounts_index.zero_lamport_pubkeys().clone();
+        timings.zero_lamport_count = pubkeys.len() as u64;
+        zero_lamport_key_clone.stop();
+        timings.zero_lamport_key_clone_us += zero_lamport_key_clone.as_us();
+
+        let mut collect_delta_keys = Measure::start("key_create");
+        let max_slot = max_clean_root.unwrap_or_else(|| self.accounts_index.max_root());
+        let (delta_keys, _max_slot) = self.collect_uncleaned_pubkeys_to_slot(max_slot);
+        collect_delta_keys.stop();
+        timings.collect_delta_keys_us += collect_delta_keys.as_us();
+
+        let mut delta_insert = Measure::start("delta_insert");
+        self.thread_pool_clean.install(|| {
+            delta_keys.par_iter().for_each(|keys| {
+                for key in keys {
+                    pubkeys.insert(*key);
+                }
+            });
+        });
+
+        {
+            let mut squashed_uncleaned_pubkeys = self.squashed_uncleaned_pubkeys.write().unwrap();
+            if squashed_uncleaned_pubkeys.1 <= max_slot {
```

Comment:
@ryoqun, @sakridge A few thoughts I had while rebasing https://github.com/solana-labs/solana/pull/14596. 

1) Since `max_slot` is constrained by the current scan root, but the cache is always flushing all roots (so `squashed_uncleaned_pubkeys.1` is progressing regardless of the scan root) , is it possible that continuous scans could constantly trigger this condition and prevent clean from happening()? I was thinking maybe we need like a `Pubkey -> Slot` map to see which pubkeys are available for cleaning.

2) Since `squashed_uncleaned_pubkeys` is almost always is *only* updated on `force_flush_accounts_cache()` of the roots, which happens before clean_accounts, is there any reason to squash in the flush instead of here? 

One  reason I could think of is in the exceptional case where `flush_accounts_cache_if_needed()`  flushes the roots because more than  the limit number of slots in the cache, and we might want to condense memory usage there. But then of course problem 1) above is still an issue.

3) https://github.com/solana-labs/solana/pull/14652/files#diff-1090394420d51617f3233275c2b65ed706b35b53b115fe65f82c682af8134a6fR3567-R3570 I think this may also be able to be moved from blocking `freeze()` into `flush_accounts_cache()` as well, since we know the pubkeys at flushing time as well, and can do some redundant cleaning based on later slots in the cache (can be added to my PR: https://github.com/solana-labs/solana/pull/14596).

Reply:
1. Yea, I think it's possible it prevents it. Another option is just to be more conservative with what root it squashes to. Maybe squash in batches of 10-50 slots.
2. Well if we only squash right before `clean_accounts`, we don't need `squashed_uncleaned_pubkeys` at all. We just coalesce from the `uncleaned_pubkeys`. The squash from `flush_accounts_if_needed` was to lower the memory usage of the dirty key set by coalescing multiple slots.
3. Yea, that sounds good.